borsh-rs/borsh/src/de/mod.rs

```diff
@@ -22,25 +24,98 @@ pub trait BorshDeserialize: Sized {
         }
         Ok(result)
     }
+
+    /// Returns the size in bytes of Self, if the size is fixed without dynamic memory inside,
+    /// e.g. u32 -> Some(4)
+    fn get_fixed_size() -> Option<usize> {
+        None
+    }
+
+    /// Whether Self is u8.
+    /// NOTE: `Vec<u8>` is the most common use-case for serialization and deserialization, it's
+    /// worth handling it as a special case to improve performance.
+    /// It's a workaround for specific `Vec<u8>` implementation versus generic `Vec<T>`
+    /// implementation. See https://github.com/rust-lang/rfcs/pull/1210 for details.
+    #[inline]
+    fn is_u8() -> bool {
+        false
+    }
 }
 
 impl BorshDeserialize for u8 {
     #[inline]
-    fn deserialize<R: Read>(reader: &mut R) -> Result<Self, Error> {
-        let mut res = 0u8;
-        reader.read_exact(std::slice::from_mut(&mut res))?;
+    fn deserialize<R: BufRead>(reader: &mut R) -> Result<Self, Error> {
+        let res = {
+            let buf = reader.fill_buf()?;
+            if buf.is_empty() {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::InvalidInput,
+                    ERROR_UNEXPECTED_LENGTH_OF_INPUT,
+                ));
+            }
+            buf[0]
+        };
+        reader.consume(size_of::<u8>());
         Ok(res)
     }
+
+    #[inline]
+    fn try_from_slice(v: &[u8]) -> Result<Self, Error> {
+        if v.len() == size_of::<u8>() {
+            Ok(v[0])
+        } else {
+            Err(std::io::Error::new(
+                std::io::ErrorKind::InvalidInput,
+                ERROR_UNEXPECTED_LENGTH_OF_INPUT,
+            ))
+        }
+    }
+
+    #[inline]
+    fn get_fixed_size() -> Option<usize> {
+        Some(size_of::<u8>())
+    }
+
+    #[inline]
+    fn is_u8() -> bool {
+        true
+    }
 }
 
 macro_rules! impl_for_integer {
     ($type: ident) => {
         impl BorshDeserialize for $type {
             #[inline]
-            fn deserialize<R: Read>(reader: &mut R) -> Result<Self, Error> {
-                let mut data = [0u8; size_of::<$type>()];
-                reader.read_exact(&mut data)?;
-                Ok($type::from_le_bytes(data))
+            fn deserialize<R: BufRead>(reader: &mut R) -> Result<Self, Error> {
+                let res = {
+                    let buf = reader.fill_buf()?;
+                    if buf.len() < size_of::<$type>() {
```

Comment:
`buf.len()` is the length of the remaining buffer, not the length of the actual input. It can cause several issues:
* First, the buffer length is always under the buffer capacity. By default they use 8KB, but anyone can run borsh deserializer with smaller buffer and then get suprised why this does not work;
* Just because `buf.len() < size_of::<$type>()` does not mean there is not enough data to read from the input. The buffer is populated in discrete steps and if the buffer boundary happens to be passing through the middle of some multi-byte type then we will incorrectly err. I think the reason why we don't catch it with our error messsages is because it is very unlikely though possible.

I suggest instead of trying to workaround this issue, we can use `slice` instead of `BufRead` to avoid thinking of buffering altogether.

Reply:
The only case when BufRead should return buffer with less capacity than the buffer length is when the capacity of the buffer is less than `size_of::<$type>()`. Of course, it's possible if someone constructs a `BufRead` with capacity less than 8 bytes, but it's not standard scenario. 

> I suggest instead of trying to workaround this issue, we can use `slice` instead of `BufRead` to avoid thinking of buffering altogether.

This is not simple, because we'll have to modify everything else auto-generated that uses `deserialize` right now.

Reply:
> The only case when BufRead should return buffer with less capacity than the buffer length

There is probably a typo.  BufRead capacity is always larger or equal to length.

I am not sure I communicated the problem clearly. Let me try again:
* Suppose our BufRead uses 8KB and we are deserializing a large structure (let's say 20KB):
```
struct Large {
field1: F1,
field2: F2,
...
}
```
* Upon first call to `fill_buf` BufRead will read 8KB and set length=8KB and capacity=8KB;
* Fast forward. Suppose we deserialized so far 8KB - 2 bytes, and length=2bytes, while capacity is still 8KB. We are now about to deserialize `field100: u64`;
* We call `u64::deserialize` which first calls `let buf = reader.fill_buf()?;`. But since there are still 2 bytes left it does nothing and returns a slice of 2 bytes, see: https://doc.rust-lang.org/src/std/io/buffered.rs.html#284 Here `self.pos == self.cap - 2`. We then check `buf.len() < size_of::<$type>()` which fails because `buf.len() == 2` and `size_of::<u64>() == 8`.

Reply:
I see. I misunderstood how it works. Good point.